app/utils.ts

```diff
@@ -105,6 +105,26 @@ export function useUser(): User {
   return maybeUser;
 }
 
+export async function useEffectiveUser(): Promise<User> {
```

Comment:
I think this should be:

```ts
useEffectiveUsername()
```

The `user` object will be identical.  It's just that we're going to use this hook to get the specific "username" that the current user is using, which will either be:

1. `user.effectiveUsername` if the cookie is set and we're using someone else's username
2. `user.username` if the cookie is not set and we're using our own username

```ts
export async function useEffectiveUsername(): User['username'] {
  // The user must be logged in to call this.  Use `useOptionalUser()` if you're not sure
  const { username, effectiveUsername } = useUser();
 
  // Your username is either your real username, or the username you are temporarily using as an admin 
  return effectiveUsername || username;
}
```

Reply:
I'm not sure how to go about this still, I have replaced what I had with what is suggested and it works for when we display it like on header instead of rendering a property of the user object I just render the effective username but we would have to have a different effective user object for the use of baseDomain for example. Or we could change how the method works so that getBaseDomain checks whether isAdmin and a non-empty effective username exist together and if that is the case will instead use the effectiveUsername.